File: android/feature/scripture/src/main/kotlin/com/nexussoft/verbum/feature/scripture/DailyVerseFeature.kt
package com.nexussoft.verbum.feature.scripture

import com.nexussoft.verbum.clients.BibleClient
import com.nexussoft.verbum.clients.NotificationAuthorization
import com.nexussoft.verbum.clients.NotificationClient
import com.nexussoft.verbum.clients.PreferencesClient
import com.nexussoft.verbum.clients.VerseNotification
import com.nexussoft.verbum.clients.helloao.HelloAOTranslation
import com.nexussoft.verbum.common.arch.Effect
import com.nexussoft.verbum.common.arch.Reducer
import com.nexussoft.verbum.common.arch.merge
import com.nexussoft.verbum.common.arch.only
import com.nexussoft.verbum.common.arch.runEffect
import com.nexussoft.verbum.common.arch.with
import com.nexussoft.verbum.models.BiblePassage
import com.nexussoft.verbum.models.DailyVerses
import com.nexussoft.verbum.models.PassageReference
import java.time.Clock
import java.time.LocalDate
import java.time.LocalDateTime
import kotlinx.coroutines.ensureActive

/**
 * The verse of the day on Home (§14): one verse, its text, a way to open it in the reader, a
 * way to share it, and an opt-in morning notification. Twin of iOS `DailyVerseFeature`.
 */
object DailyVerseFeature {
    data class State(
        val reference: PassageReference = DailyVerses.pool[0],
        val text: TextState = TextState.Loading,
        /** The user's choice. Kept while the system permission is denied, so re-enabling in Settings is enough. */
        val morningsEnabled: Boolean = false,
        val reminderMinute: Int = 7 * 60,
        val authorization: NotificationAuthorization = NotificationAuthorization.NOT_DETERMINED,
    ) {
        /** What the share sheet sends: the verse, its reference, its translation. */
        val shareText: String? get() = (text as? TextState.Loaded)?.let { DailyVerseShare.text(it.passage, reference) }

        /** The toggle is on and the system agrees. */
        val morningsActive: Boolean get() = morningsEnabled && authorization == NotificationAuthorization.AUTHORIZED
    }

    sealed interface TextState {
        data object Loading : TextState
        data class Loaded(val passage: BiblePassage) : TextState
        data object Failed : TextState
    }

    sealed interface Action {
        data object Started : Action
        data class TextLoaded(val passage: BiblePassage) : Action
        data object TextFailed : Action
        data class AuthorizationResponse(val authorization: NotificationAuthorization) : Action
        data object OpenTapped : Action
        data class MorningsToggled(val enabled: Boolean) : Action
        data class ReminderTimeChanged(val minuteOfDay: Int) : Action
        data class Delegate(val delegate: DelegateAction) : Action
    }

    sealed interface DelegateAction {
        data class OpenPassage(val reference: PassageReference) : DelegateAction
    }

    /** How far ahead notifications are planned; refreshed on every launch. */
    const val PLANNED_DAYS = 14
    const val MORNINGS_KEY = "dailyVerseMornings"
    const val REMINDER_MINUTE_KEY = "dailyVerseReminderMinute"
    private const val PLAN_ID = "dailyVerse.plan"

    fun reducer(
        bibleClient: BibleClient,
        notifications: NotificationClient,
        preferences: PreferencesClient,
        clock: Clock = Clock.systemDefaultZone(),
        title: () -> String = { "Verse of the day" },
    ): Reducer<State, Action> {
        /** Schedules the next [PLANNED_DAYS] mornings, replacing what was pending. Skipped without permission. */
        fun plan(minuteOfDay: Int): Effect<Action> = runEffect(id = PLAN_ID, cancelInFlight = true) {
            if (notifications.authorization() != NotificationAuthorization.AUTHORIZED) return@runEffect
            val plan = DailyVersePlan.build(LocalDateTime.now(clock), PLANNED_DAYS, minuteOfDay / 60, title(), minuteOfDay % 60) { reference ->
                runCatching { bibleClient.passage(reference).text }.getOrNull()
            }
            kotlinx.coroutines.currentCoroutineContext().ensureActive()
            notifications.scheduleVerses(plan)
        }

        return Reducer { state, action ->
            when (action) {
                Action.Started -> {
                    val enabled = preferences.string(MORNINGS_KEY) == "true"
                    val reference = DailyVerses.verse(LocalDate.now(clock))
                    val minute = preferences.string(REMINDER_MINUTE_KEY)?.toIntOrNull()?.takeIf { it in 0..1439 } ?: 7 * 60
                    state.copy(reference = reference, text = TextState.Loading, morningsEnabled = enabled, reminderMinute = minute).with(
                        merge(
                            runEffect { send ->
                                val passage = runCatching { bibleClient.passage(reference) }.getOrNull()
                                send(if (passage != null) Action.TextLoaded(passage) else Action.TextFailed)
                            },
                            runEffect { send -> send(Action.AuthorizationResponse(notifications.authorization())) },
                        ),
                    )
                }
                is Action.TextLoaded -> state.copy(text = TextState.Loaded(action.passage)).only()
                Action.TextFailed -> state.copy(text = TextState.Failed).only()
                is Action.AuthorizationResponse -> state.copy(authorization = action.authorization).with(
                    if (action.authorization == NotificationAuthorization.AUTHORIZED && state.morningsEnabled) plan(state.reminderMinute) else Effect.None,
                )
                Action.OpenTapped -> state.with(Effect.Send(Action.Delegate(DelegateAction.OpenPassage(state.reference))))
                is Action.MorningsToggled -> {
                    preferences.setString(MORNINGS_KEY, action.enabled.toString())
                    if (action.enabled) {
                        state.copy(morningsEnabled = true).with(
                            runEffect { send ->
                                val granted = runCatching { notifications.requestAuthorization() }.getOrDefault(false)
                                send(Action.AuthorizationResponse(if (granted) NotificationAuthorization.AUTHORIZED else NotificationAuthorization.DENIED))
                            },
                        )
                    } else {
                        state.copy(morningsEnabled = false).with(merge(
                            runEffect(id = PLAN_ID, cancelInFlight = true) {},
                            runEffect { notifications.cancelVerses() },
                        ))
                    }
                }
                is Action.ReminderTimeChanged -> {
                    if (action.minuteOfDay !in 0..1439) state.only()
                    else {
                        preferences.setString(REMINDER_MINUTE_KEY, action.minuteOfDay.toString())
                        state.copy(reminderMinute = action.minuteOfDay).with(
                            if (state.morningsEnabled) plan(action.minuteOfDay) else Effect.None,
                        )
                    }
                }
                is Action.Delegate -> state.only()
            }
        }
    }
}

/** Builds the dated notifications for the mornings ahead. Pure apart from [text]. Twin of iOS `DailyVersePlan`. */
object DailyVersePlan {
    suspend fun build(now: LocalDateTime, days: Int, hour: Int, title: String, minute: Int = 0, text: suspend (PassageReference) -> String?): List<VerseNotification> {
        // Today's morning only if it is still ahead; otherwise the plan starts tomorrow.
        val first = if (now.hour * 60 + now.minute < hour * 60 + minute) now.toLocalDate() else now.toLocalDate().plusDays(1)
        return (0 until days).map { offset ->
            val date = first.plusDays(offset.toLong())
            val reference = DailyVerses.verse(date)
            VerseNotification(
                year = date.year, month = date.monthValue, day = date.dayOfMonth, hour = hour, minute = minute,
                title = title,
                body = body(reference, text(reference)),
                reference = reference,
            )
        }
    }

    /** `“text” — John 3:16`, or just the reference when the text could not be read while planning. */
    fun body(reference: PassageReference, text: String?): String =
        if (text.isNullOrEmpty()) reference.formatted else "“$text” — ${reference.formatted}"
}

/** What leaves the app when the verse is shared. Twin of iOS `DailyVerseShare`. */
object DailyVerseShare {
    fun text(passage: BiblePassage, reference: PassageReference): String =
        "“${passage.text}”\n— ${reference.formatted} · ${HelloAOTranslation.name(passage.translationId)}"
}

File: VerbumKit/Sources/Features/DailyVerse/DailyVerseFeature.swift
import Clients
import ComposableArchitecture
import Foundation
import Models

/// The verse of the day on Home (spec §14): one verse, its text, a way to open
/// it in the reader, a way to share it, and an opt-in morning notification.
/// Every device shows the same verse on the same day (`DailyVerses`).
@Reducer
public struct DailyVerseFeature {
    @ObservableState
    public struct State: Equatable {
        public var reference: PassageReference = DailyVerses.pool[0]
        public var text: TextState = .loading
        /// The user's choice. Kept even while the system permission is denied,
        /// so turning notifications back on in Settings is enough.
        @Shared(.dailyVerseMornings) public var morningsEnabled
        @Shared(.dailyVerseReminderMinute) public var reminderMinute
        public var authorization: NotificationAuthorization = .notDetermined

        public init() {}

        /// What the share sheet sends: the verse, its reference, its translation.
        public var shareText: String? {
            guard case .loaded(let passage) = text else { return nil }
            return DailyVerseShare.text(passage, reference: reference)
        }

        /// The toggle is on and the system agrees.
        public var morningsActive: Bool { morningsEnabled && authorization == .authorized }
    }

    public enum TextState: Equatable, Sendable {
        case loading
        case loaded(BiblePassage)
        case failed
    }

    public enum Action: Equatable {
        case task
        case textResponse(Result<BiblePassage, Failure>)
        case authorizationResponse(NotificationAuthorization)
        case openTapped
        case morningsToggled(Bool)
        case reminderTimeChanged(Int)
        case delegate(Delegate)

        @CasePathable
        public enum Delegate: Equatable {
            case openPassage(PassageReference)
        }
    }

    public struct Failure: Error, Equatable, Sendable {
        public init() {}
    }

    /// How far ahead notifications are planned; refreshed on every launch.
    public static let plannedDays = 14

    @Dependency(\.date.now) var now
    @Dependency(\.calendar) var calendar
    @Dependency(\.bibleClient) var bibleClient
    @Dependency(\.notificationClient) var notificationClient

    public init() {}

    public var body: some ReducerOf<Self> {
        Reduce { state, action in
            switch action {
            case .task:
                state.reference = DailyVerses.verse(on: now, calendar: calendar)
                state.text = .loading
                return .merge(
                    .run { [bibleClient, reference = state.reference] send in
                        do {
                            await send(.textResponse(.success(try await bibleClient.passage(reference))))
                        } catch {
                            await send(.textResponse(.failure(Failure())))
                        }
                    },
                    .run { [notificationClient] send in
                        await send(.authorizationResponse(await notificationClient.authorization()))
                    }
                )

            case .textResponse(.success(let passage)):
                state.text = .loaded(passage)
                return .none

            case .textResponse(.failure):
                state.text = .failed
                return .none

            case .authorizationResponse(let authorization):
                state.authorization = authorization
                return authorization == .authorized && state.morningsEnabled
                    ? plan(minuteOfDay: state.reminderMinute) : .none

            case .openTapped:
                return .send(.delegate(.openPassage(state.reference)))

            case .morningsToggled(true):
                state.$morningsEnabled.withLock { $0 = true }
                return .run { [notificationClient] send in
                    let granted = (try? await notificationClient.requestAuthorization()) ?? false
                    await send(.authorizationResponse(granted ? .authorized : .denied))
                }

            case .morningsToggled(false):
                state.$morningsEnabled.withLock { $0 = false }
                return .concatenate(.cancel(id: CancelID.plan), .run { [notificationClient] _ in await notificationClient.cancelVerses() })

            case .reminderTimeChanged(let minute):
                guard (0..<1440).contains(minute) else { return .none }
                state.$reminderMinute.withLock { $0 = minute }
                return state.morningsEnabled ? plan(minuteOfDay: minute) : .none

            case .delegate:
                return .none
            }
        }
    }

    /// Schedules the next `plannedDays` mornings, replacing what was pending.
    /// Skipped silently when the user has not granted notifications: asking
    /// the system to schedule would be a no-op, and the plan is rebuilt on the
    /// next launch anyway.
    private func plan(minuteOfDay: Int) -> Effect<Action> {
        .run { [now, calendar, bibleClient, notificationClient] _ in
            guard await notificationClient.authorization() == .authorized else { return }
            let notifications = await DailyVersePlan.build(
                from: now,
                calendar: calendar,
                days: Self.plannedDays,
                hour: minuteOfDay / 60,
                minute: minuteOfDay % 60,
                title: L10n.t("Verse of the day"),
                text: { reference in try? await bibleClient.passage(reference).text }
            )
            guard !Task.isCancelled else { return }
            await notificationClient.scheduleVerses(notifications)
        }
        .cancellable(id: CancelID.plan, cancelInFlight: true)
    }

    private enum CancelID { case plan }
}

extension SharedKey where Self == AppStorageKey<Int>.Default {
    public static var dailyVerseReminderMinute: Self {
        Self[.appStorage("dailyVerseReminderMinute"), default: 7 * 60]
    }
}

extension SharedKey where Self == AppStorageKey<Bool>.Default {
    /// Whether the user asked for the verse every morning.
    public static var dailyVerseMornings: Self {
        Self[.appStorage("dailyVerseMornings"), default: false]
    }
}

/// Builds the dated notifications for the mornings ahead. Pure apart from
/// `text`, so tests can pin the dates and the fallback body.
public enum DailyVersePlan {
    public static func build(
        from now: Date,
        calendar: Calendar,
        days: Int,
        hour: Int,
        minute: Int = 0,
        title: String,
        text: @Sendable (PassageReference) async -> String?
    ) async -> [VerseNotification] {
        var notifications: [VerseNotification] = []
        let today = calendar.startOfDay(for: now)
        // The chosen local time today is included only while still ahead.
        let nowMinute = calendar.component(.hour, from: now) * 60 + calendar.component(.minute, from: now)
        let firstOffset = nowMinute < hour * 60 + minute ? 0 : 1
        for offset in firstOffset..<(firstOffset + days) {
            guard let date = calendar.date(byAdding: .day, value: offset, to: today) else { continue }
            let c = calendar.dateComponents([.year, .month, .day], from: date)
            guard let year = c.year, let month = c.month, let day = c.day else { continue }
            let reference = DailyVerses.verse(year: year, month: month, day: day)
            let text = await text(reference)
            notifications.append(VerseNotification(
                year: year, month: month, day: day, hour: hour, minute: minute,
                title: title,
                body: body(reference: reference, text: text),
                reference: reference
            ))
        }
        return notifications
    }

    /// `“text” — John 3:16`, or just the reference when the text could not be
    /// read while planning (the tap still opens the verse).
    static func body(reference: PassageReference, text: String?) -> String {
        guard let text, !text.isEmpty else { return reference.formatted }
        return "\u{201C}\(text)\u{201D} — \(reference.formatted)"
    }
}

/// What leaves the app when the verse is shared.
public enum DailyVerseShare {
    public static func text(_ passage: BiblePassage, reference: PassageReference) -> String {
        "\u{201C}\(passage.text)\u{201D}\n— \(reference.formatted) · \(HelloAOTranslation.name(for: passage.translationId))"
    }
}
